sdk/containerregistry/Azure.Containers.ContainerRegistry/src/Authentication/ContainerRegistryCredentialsPolicy.cs

```diff
@@ -0,0 +1,144 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Collections.Generic;
+using System.Threading.Tasks;
+using Azure.Core;
+using Azure.Core.Pipeline;
+
+namespace Azure.Containers.ContainerRegistry
+{
+    /// <summary>
+    /// Challenge-based authentication policy for Container Registry Service.
+    ///
+    /// The challenge-based authorization flow is illustrated in the following steps.
+    /// For example, GET /api/v1/acr/repositories translates into the following calls.
+    ///
+    /// Step 1: GET /api/v1/acr/repositories
+    /// Return Header: 401: www-authenticate header - Bearer realm="{url}",service="{serviceName}",scope="{scope}",error="invalid_token"
+    ///
+    /// Step 2: Retrieve the serviceName, scope from the WWW-Authenticate header.  (Parse the string.)
+    ///
+    /// Step 3: POST /api/oauth2/exchange
+    /// Request Body : { service, scope, grant-type, aadToken with ARM scope }
+    /// Response Body: { acrRefreshToken }
+    ///
+    /// Step 4: POST /api/oauth2/token
+    /// Request Body: { acrRefreshToken, scope, grant-type }
+    /// Response Body: { acrAccessToken }
+    ///
+    /// Step 5: GET /api/v1/acr/repositories
+    /// Request Header: { Bearer acrTokenAccess }
+    /// </summary>
+    internal class ContainerRegistryCredentialsPolicy : BearerTokenChallengeAuthenticationPolicy
+    {
+        private readonly RefreshTokensRestClient _exchangeRestClient;
+        private readonly AccessTokensRestClient _tokenRestClient;
+
+        public ContainerRegistryCredentialsPolicy(TokenCredential credential, string aadScope, RefreshTokensRestClient exchangeRestClient, AccessTokensRestClient tokenRestClient)
+            : base(credential, aadScope)
+        {
+            Argument.AssertNotNull(credential, nameof(credential));
+            Argument.AssertNotNull(aadScope, nameof(aadScope));
+
+            _exchangeRestClient = exchangeRestClient;
+            _tokenRestClient = tokenRestClient;
+        }
+
+        protected override async ValueTask<bool> AuthenticateRequestOnChallengeAsync(HttpMessage message, bool async)
+        {
+            // Once we're here, we've completed Step 1.
+
+            // Step 2: Parse challenge string to retrieve serviceName and scope, where scope is the ACR Scope
+            Dictionary<string, string> challengeValues = ParseChallengeValues(message);
```

Comment:
I *think* you could replace this with calls to  `AuthorizationChallengeParser.GetChallengeParameterFromResponse`

for example:
```C#
var serviceChallenge = AuthorizationChallengeParser.GetChallengeParameterFromResponse(message.Response, "Bearer", "service");
var scopeChallenge = AuthorizationChallengeParser.GetChallengeParameterFromResponse(message.Response, "Bearer", "scope");
```

Reply:
Let me try that - if that works, it's way more efficient than the initial parser I wrote!

Reply:
Brilliant.